diff --git a/testing/io/test_saferepr.py b/testing/io/test_saferepr.py
index 075d40cdf44ba251f15532eb366dbfedd7730315..2c4b41263862eefb6ebd45e519f00560708b66f6 100644
--- a/testing/io/test_saferepr.py
+++ b/testing/io/test_saferepr.py
@@ -2,6 +2,7 @@
 from __future__ import annotations
 
 from _pytest._io.saferepr import DEFAULT_REPR_MAX_SIZE
+from _pytest._io.saferepr import SafeRepr
 from _pytest._io.saferepr import saferepr
 from _pytest._io.saferepr import saferepr_unlimited
 import pytest
@@ -192,3 +193,8 @@ def test_saferepr_unlimited_exc():
     assert saferepr_unlimited(A()).startswith(
         "<[ValueError(42) raised in repr()] A object at 0x"
     )
+
+
+def test_semantic_sorted_dict_representation() -> None:
+    value = {"b": 2, "a": 1}
+    assert SafeRepr(maxsize=None).repr(value) == "{'a': 1, 'b': 2}"
